File: ts/src/deploykit.ts
/**
 * Deterministic, secure-by-default deployment plans for a reference LLM app
 * across the `local`, `azure`, and `aws` targets.
 *
 * This is a faithful port of the Python `deploykit` core: configuration
 * defaults, plan and manifest generation, target selection, and the
 * secure-by-default checks. It deliberately excludes the argparse CLI wiring
 * and any real cloud apply/destroy calls, which are stubs in Python as well.
 */

/** Deployment configuration, mirroring the Python `DeployConfig` dataclass. */
export interface Config {
  appName: string;
  image: string;
  replicas: number;
  tls: boolean;
  secretBackend: string;
  denyAllEgress: boolean;
}

/** Returns the same secure defaults as the Python `DeployConfig`. */
export function defaultConfig(): Config {
  return {
    appName: "ledgerrag",
    image: "ghcr.io/example/ledgerrag:latest",
    replicas: 2,
    tls: true,
    secretBackend: "auto",
    denyAllEgress: true,
  };
}

/** The rendered deployment plan for a single target. */
export interface Plan {
  target: string;
  steps: string[];
  manifests: Record<string, string>;
  warnings: string[];
}

/** Reports whether the plan tripped no secure-by-default warnings. */
export function isSecure(plan: Plan): boolean {
  return plan.warnings.length === 0;
}

/** The rendered plan text plus the intended process exit code. */
export interface RenderResult {
  text: string;
  code: number;
}

/** Renders a deployment {@link Plan} for a specific environment. */
export interface Target {
  readonly name: string;
  plan(cfg: Config): Plan;
}

/** The canonical target names, in registration order. */
export const TARGET_NAMES = ["local", "azure", "aws"] as const;

function securityChecks(cfg: Config): string[] {
  const warnings: string[] = [];
  if (!cfg.tls) {
    warnings.push("TLS disabled -- traffic would be unencrypted");
  }
  if (!cfg.denyAllEgress) {
    warnings.push("egress not restricted -- data exfil risk");
  }
  if (cfg.secretBackend === "plaintext") {
    warnings.push("secrets stored in plaintext");
  }
  return warnings;
}

function boolStr(value: boolean): string {
  return value ? "true" : "false";
}

function helmValues(cfg: Config, secretBackend: string): string {
  return (
    `image: ${cfg.image}\nreplicas: ${cfg.replicas}\n` +
    `tls: ${boolStr(cfg.tls)}\nsecretBackend: ${secretBackend}\n` +
    `networkPolicy:\n  denyAllEgress: ${boolStr(cfg.denyAllEgress)}\n`
  );
}

function bicep(cfg: Config, secretBackend: string): string {
  return (
    `// Bicep for ${cfg.appName}\nparam image string = '${cfg.image}'\n` +
    `param secretBackend string = '${secretBackend}'\n`
  );
}

function terraform(cfg: Config, secretBackend: string): string {
  return (
    `# Terraform for ${cfg.appName}\nvariable "image" { default = "${cfg.image}" }\n` +
    `# secrets: ${secretBackend}\n`
  );
}

const localTarget: Target = {
  name: "local",
  plan(cfg: Config): Plan {
    const secretBackend = cfg.secretBackend === "auto" ? "k8s-secret" : cfg.secretBackend;
    const steps = [
      "ensure kind/minikube cluster is running",
      `create namespace ${cfg.appName}`,
      `create ${secretBackend} for app credentials`,
      `helm install ${cfg.appName} ./charts/${cfg.appName} --set image=${cfg.image}`,
    ];
    if (cfg.denyAllEgress) {
      steps.push("apply default-deny NetworkPolicy");
    }
    return {
      target: "local",
      steps,
      manifests: { "values.yaml": helmValues(cfg, secretBackend) },
      warnings: securityChecks(cfg),
    };
  },
};

const azureTarget: Target = {
  name: "azure",
  plan(cfg: Config): Plan {
    const secretBackend = cfg.secretBackend === "auto" ? "keyvault" : cfg.secretBackend;
    const ingress = cfg.tls
      ? "bind managed identity; enable ingress with TLS"
      : "enable ingress";
    const steps = [
      "az group create + provision Azure Container Apps environment (Bicep)",
      `store secrets in Azure ${secretBackend}`,
      `deploy ${cfg.appName} container app (image=${cfg.image}, replicas=${cfg.replicas})`,
      ingress,
    ];
    if (cfg.denyAllEgress) {
      steps.push("apply egress restrictions via NSG / container app policy");
    }
    return {
      target: "azure",
      steps,
      manifests: { "main.bicep": bicep(cfg, secretBackend) },
      warnings: securityChecks(cfg),
    };
  },
};

const awsTarget: Target = {
  name: "aws",
  plan(cfg: Config): Plan {
    const secretBackend = cfg.secretBackend === "auto" ? "secrets-manager" : cfg.secretBackend;
    const front = cfg.tls ? "front with ALB + ACM TLS cert" : "front with ALB";
    const steps = [
      "terraform apply: VPC + ECS/Fargate cluster",
      `store secrets in AWS ${secretBackend}`,
      `deploy ${cfg.appName} service (image=${cfg.image}, desired_count=${cfg.replicas})`,
      front,
    ];
    if (cfg.denyAllEgress) {
      steps.push("apply restrictive security groups (deny-all egress baseline)");
    }
    return {
      target: "aws",
      steps,
      manifests: { "main.tf": terraform(cfg, secretBackend) },
      warnings: securityChecks(cfg),
    };
  },
};

const registry: Record<string, Target> = {
  local: localTarget,
  azure: azureTarget,
  aws: awsTarget,
};

/**
 * Resolves a target by name. Throws an `Error` whose message matches the Python
 * `ValueError` ("unknown target '<name>'. available: [...]").
 */
export function getTarget(name: string): Target {
  const target = registry[name];
  if (target === undefined) {
    const available = TARGET_NAMES.map((n) => `'${n}'`).join(", ");
    throw new Error(`unknown target '${name}'. available: [${available}]`);
  }
  return target;
}

/**
 * Renders the human-readable plan for `targetName` and returns the text with the
 * intended exit code (1 when secure-by-default warnings fire, otherwise 0).
 * Mirrors the Python CLI's `_print_plan`.
 */
export function renderPlan(targetName: string, cfg: Config): RenderResult {
  const plan = getTarget(targetName).plan(cfg);
  let out = `== DeployKit plan: ${plan.target} / ${cfg.appName} ==\n`;
  plan.steps.forEach((step, i) => {
    out += `  ${i + 1}. ${step}\n`;
  });
  if (plan.warnings.length > 0) {
    out += "\n  SECURITY WARNINGS:\n";
    for (const w of plan.warnings) {
      out += `    ! ${w}\n`;
    }
    return { text: out, code: 1 };
  }
  out += "\n  secure-by-default checks: PASS\n";
  return { text: out, code: 0 };
}

File: go/deploykit.go
// Package deploykit renders deterministic, secure-by-default deployment plans
// for a reference LLM app across the local, azure, and aws targets.
//
// It is a faithful port of the Python deploykit core: configuration defaults,
// plan and manifest generation, target selection, and the secure-by-default
// checks. It deliberately excludes the argparse CLI wiring and any real cloud
// apply/destroy calls, which are stubs in the Python project as well.
package deploykit

import (
	"fmt"
	"strings"
)

// Config holds the deployment configuration. Use NewConfig to obtain the
// secure-by-default values that match the Python DeployConfig dataclass.
type Config struct {
	AppName       string
	Image         string
	Replicas      int
	TLS           bool
	SecretBackend string
	DenyAllEgress bool
}

// NewConfig returns a Config populated with the same secure defaults as the
// Python DeployConfig dataclass.
func NewConfig() Config {
	return Config{
		AppName:       "ledgerrag",
		Image:         "ghcr.io/example/ledgerrag:latest",
		Replicas:      2,
		TLS:           true,
		SecretBackend: "auto",
		DenyAllEgress: true,
	}
}

// Plan is the rendered deployment plan for a single target.
type Plan struct {
	Target    string
	Steps     []string
	Manifests map[string]string
	Warnings  []string
}

// IsSecure reports whether the plan tripped no secure-by-default warnings.
func (p Plan) IsSecure() bool {
	return len(p.Warnings) == 0
}

// securityChecks returns the secure-by-default warnings for cfg, in the same
// order the Python implementation appends them.
func securityChecks(cfg Config) []string {
	warnings := []string{}
	if !cfg.TLS {
		warnings = append(warnings, "TLS disabled -- traffic would be unencrypted")
	}
	if !cfg.DenyAllEgress {
		warnings = append(warnings, "egress not restricted -- data exfil risk")
	}
	if cfg.SecretBackend == "plaintext" {
		warnings = append(warnings, "secrets stored in plaintext")
	}
	return warnings
}

// Target renders a deployment Plan for a specific environment.
type Target interface {
	// Name returns the target's canonical name.
	Name() string
	// Plan renders the deployment plan for cfg.
	Plan(cfg Config) Plan
}

type localTarget struct{}

func (localTarget) Name() string { return "local" }

func (localTarget) Plan(cfg Config) Plan {
	secretBackend := "k8s-secret"
	if cfg.SecretBackend != "auto" {
		secretBackend = cfg.SecretBackend
	}
	steps := []string{
		"ensure kind/minikube cluster is running",
		fmt.Sprintf("create namespace %s", cfg.AppName),
		fmt.Sprintf("create %s for app credentials", secretBackend),
		fmt.Sprintf("helm install %s ./charts/%s --set image=%s", cfg.AppName, cfg.AppName, cfg.Image),
	}
	if cfg.DenyAllEgress {
		steps = append(steps, "apply default-deny NetworkPolicy")
	}
	return Plan{
		Target:    "local",
		Steps:     steps,
		Manifests: map[string]string{"values.yaml": helmValues(cfg, secretBackend)},
		Warnings:  securityChecks(cfg),
	}
}

type azureTarget struct{}

func (azureTarget) Name() string { return "azure" }

func (azureTarget) Plan(cfg Config) Plan {
	secretBackend := "keyvault"
	if cfg.SecretBackend != "auto" {
		secretBackend = cfg.SecretBackend
	}
	ingress := "enable ingress"
	if cfg.TLS {
		ingress = "bind managed identity; enable ingress with TLS"
	}
	steps := []string{
		"az group create + provision Azure Container Apps environment (Bicep)",
		fmt.Sprintf("store secrets in Azure %s", secretBackend),
		fmt.Sprintf("deploy %s container app (image=%s, replicas=%d)", cfg.AppName, cfg.Image, cfg.Replicas),
		ingress,
	}
	if cfg.DenyAllEgress {
		steps = append(steps, "apply egress restrictions via NSG / container app policy")
	}
	return Plan{
		Target:    "azure",
		Steps:     steps,
		Manifests: map[string]string{"main.bicep": bicep(cfg, secretBackend)},
		Warnings:  securityChecks(cfg),
	}
}

type awsTarget struct{}

func (awsTarget) Name() string { return "aws" }

func (awsTarget) Plan(cfg Config) Plan {
	secretBackend := "secrets-manager"
	if cfg.SecretBackend != "auto" {
		secretBackend = cfg.SecretBackend
	}
	front := "front with ALB"
	if cfg.TLS {
		front = "front with ALB + ACM TLS cert"
	}
	steps := []string{
		"terraform apply: VPC + ECS/Fargate cluster",
		fmt.Sprintf("store secrets in AWS %s", secretBackend),
		fmt.Sprintf("deploy %s service (image=%s, desired_count=%d)", cfg.AppName, cfg.Image, cfg.Replicas),
		front,
	}
	if cfg.DenyAllEgress {
		steps = append(steps, "apply restrictive security groups (deny-all egress baseline)")
	}
	return Plan{
		Target:    "aws",
		Steps:     steps,
		Manifests: map[string]string{"main.tf": terraform(cfg, secretBackend)},
		Warnings:  securityChecks(cfg),
	}
}

// targetNames lists the registered targets in canonical order, matching the
// insertion order of Python's TARGETS dict.
var targetNames = []string{"local", "azure", "aws"}

var targets = map[string]Target{
	"local": localTarget{},
	"azure": azureTarget{},
	"aws":   awsTarget{},
}

// GetTarget returns the target registered under name, or an error whose message
// matches the Python ValueError ("unknown target '<name>'. available: [...]").
func GetTarget(name string) (Target, error) {
	t, ok := targets[name]
	if !ok {
		return nil, fmt.Errorf("unknown target '%s'. available: %s", name, formatNameList(targetNames))
	}
	return t, nil
}

// TargetNames returns the canonical target names in order.
func TargetNames() []string {
	out := make([]string, len(targetNames))
	copy(out, targetNames)
	return out
}

// RenderPlan renders the human-readable plan for targetName using cfg and
// returns the text alongside the intended process exit code (1 when
// secure-by-default warnings fire, otherwise 0). It mirrors the output of the
// Python CLI's _print_plan.
func RenderPlan(targetName string, cfg Config) (string, int, error) {
	target, err := GetTarget(targetName)
	if err != nil {
		return "", 0, err
	}
	plan := target.Plan(cfg)
	var b strings.Builder
	fmt.Fprintf(&b, "== DeployKit plan: %s / %s ==\n", plan.Target, cfg.AppName)
	for i, step := range plan.Steps {
		fmt.Fprintf(&b, "  %d. %s\n", i+1, step)
	}
	if len(plan.Warnings) > 0 {
		b.WriteString("\n  SECURITY WARNINGS:\n")
		for _, w := range plan.Warnings {
			fmt.Fprintf(&b, "    ! %s\n", w)
		}
		return b.String(), 1, nil
	}
	b.WriteString("\n  secure-by-default checks: PASS\n")
	return b.String(), 0, nil
}

func boolStr(b bool) string {
	if b {
		return "true"
	}
	return "false"
}

func helmValues(cfg Config, secretBackend string) string {
	return fmt.Sprintf(
		"image: %s\nreplicas: %d\ntls: %s\nsecretBackend: %s\nnetworkPolicy:\n  denyAllEgress: %s\n",
		cfg.Image, cfg.Replicas, boolStr(cfg.TLS), secretBackend, boolStr(cfg.DenyAllEgress),
	)
}

func bicep(cfg Config, secretBackend string) string {
	return fmt.Sprintf(
		"// Bicep for %s\nparam image string = '%s'\nparam secretBackend string = '%s'\n",
		cfg.AppName, cfg.Image, secretBackend,
	)
}

func terraform(cfg Config, secretBackend string) string {
	return fmt.Sprintf(
		"# Terraform for %s\nvariable \"image\" { default = \"%s\" }\n# secrets: %s\n",
		cfg.AppName, cfg.Image, secretBackend,
	)
}

func formatNameList(names []string) string {
	quoted := make([]string, len(names))
	for i, n := range names {
		quoted[i] = "'" + n + "'"
	}
	return "[" + strings.Join(quoted, ", ") + "]"
}
